onnxruntime/core/providers/cpu/ml/tree_ensemble_aggregator.h

```diff
@@ -23,9 +23,20 @@ struct TreeNodeElementId {
   }
   struct hash_fn {
     std::size_t operator()(const TreeNodeElementId& key) const {
-      std::size_t h1 = std::hash<int64_t>()(key.tree_id);
-      std::size_t h2 = std::hash<int64_t>()(key.node_id);
-      return h1 ^ h2;
```

Comment:
My understanding is that the large number of collisions came from the fact that this is symmetrical. Doesn't everything just work if we do `h1 ^ (h2 <<1)`?

Reply:
I suspect the root cause here is related with the fact that both id's (tree_id's and node_id's) for a given TreeEnsembleClassifier node are chosen sequentially during the conversion to ONNX. Also, for node_id's we can have the same value being reused across different trees ("Ids may restart at zero for each tree, but it not required to" from [nodes_nodeids attribute description](https://github.com/onnx/onnx/blob/main/docs/Operators-ml.md#aionnxmltreeensembleclassifier)). I confirmed that this is happening for our model, i.e. there are repetitions of values on nodes_nodeids array.

In that case, the difference between the hash of each element of the pair end up being of few bits. After applying the XOR operator, since few bits are different, the number of collisions is high. I was able to find some examples (tables below).

Since we don't have many guarantees on the values chosen for tree_id's and node_id's (besides each pair being unique), I believe it's hard to beat Elegant Pairing in terms of collision avoidance. Even though Elegant Pairing assumes two positive integers, the edge case handling showed pretty good results.

## Using (h1 ^ h2)
All have the same hash value equals to 0xF:
```
  1 E F | 10 1F F | 20 2F F |     | 50 5F F                               
  2 D F | 11 1E F | 21 2E F |     | 51 5E F                               
  3 C F | 12 1D F | 22 2D F |     | 52 5D F                               
  4 B F | 13 1C F | 23 2C F |     | 53 5C F                               
  5 A F | 14 1B F | 24 2B F |     | 54 5B F                               
  6 9 F | 15 1A F | 25 2A F |     | 55 5A F                               
  7 8 F | 16 19 F | 26 29 F |     | 56 59 F                               
  8 7 F | 17 18 F | 27 28 F |     | 57 58 F                               
  9 6 F | 18 17 F | 28 27 F | ... | 58 57 F                               
  A 5 F | 19 16 F | 29 26 F |     | 59 56 F                               
  B 4 F | 1A 15 F | 2A 25 F |     | 5A 55 F                               
  C 3 F | 1B 14 F | 2B 24 F |     | 5B 54 F                               
  D 2 F | 1C 13 F | 2C 23 F |     | 5C 53 F                               
  E 1 F | 1D 12 F | 2D 22 F |     | 5D 52 F                               
        | 1E 11 F | 2E 21 F |     | 5E 51 F                               
        | 1F 10 F | 2F 20 F |     | 5F 50 F
```

Legend:
- 1st column: the hash value for tree_id (starting from 1)
- 2nd column: the hash value for the node_id (starting from 1)
- 3rd column: (1st ^ 2nd)

## Using (h1 ^ (h2 << 1))
All have the same hash value equals to 0x1:
```
 3 1 1 | 13 9 1 | 23 11 1 | 33 19 1 | 43 21 1 |     | 3DB 1ED 1 |
 7 3 1 | 17 B 1 | 27 13 1 | 37 1B 1 | 47 23 1 | ... | 3DF 1EF 1 |
 B 5 1 | 1B D 1 | 2B 15 1 | 3B 1D 1 | 4B 25 1 |     | 3E3 1F1 1 |
 F 7 1 | 1F F 1 | 2F 17 1 | 3F 1F 1 | 4F 27 1 |     | 3E7 1F3 1 |
```
Legend:
- 1st column: the hash value for tree_id (starting from 1)
- 2nd column: the hash value for the node_id (starting from 1)
- 3rd column: (1st ^ (2nd << 1))

Reply:
For the sake of completeness, I've repeated the experiments for the variation you proposed `h1 ^ (h2 << 1)`.

## [-10000, 10000] Interval
```
ID's pair count: 400040001
distinct hash values produced: 57344
collision count: 400040001

there are 16384 buckets with 10001 elements
there are 16384 buckets with 10000 elements
there are 16384 buckets with 3617 elements
there are 2048 buckets with 1809 elements
there are 2048 buckets with 1808 elements
...
there are 512 buckets with 1569 elements
there are 415 buckets with 1536 elements
there are 33 buckets with 1537 elements
there are 32 buckets with 1553 elements
there are 32 buckets with 1552 elements
printed first and last 5 buckets with more than one element (still 2 remaining of 12)
```

## [2^60, 2^60 + 5000] Interval

```
ID's pair count: 25010001
distinct hash values produced: 14336
collision count: 25010001

there are 4096 buckets with 2501 elements
there are 4096 buckets with 2500 elements
there are 4096 buckets with 905 elements
there are 512 buckets with 453 elements
there are 512 buckets with 452 elements
...
there are 128 buckets with 393 elements
there are 103 buckets with 384 elements
there are 9 buckets with 385 elements
there are 8 buckets with 389 elements
there are 8 buckets with 388 elements
printed first and last 5 buckets with more than one element (still 2 remaining of 12)
```

Reply:
Thanks again for the tests! This (hashing tuples without a lot of collisions) should be a solved problem; C++ might just make it hard to find. `onnxruntime` already depends on abseil in various places. We could consider using their `Hash` (which also works for tuples) here. The produced hashes seem nicely distributed: https://godbolt.org/z/qGsK6jj9a

Somewhat unrelated, but it might actually be worth generally using the abseil's hashmap rather than the `unordered_map` of the standard library. This won't save us from defining a hash function for `TreeNodeElementId`, though.

Reply:
Oh! That is true, I missed it.

I've changed the code accordingly. The results seem equivalent. 

```
$> time ./onnx_test_runner -v ~/folder_with_model
result:
	Models: 1
	Total test cases: 0
		Succeeded: 0
		Not implemented: 0
		Failed: 0
	Stats by Operator type:
		Not implemented(0):
		Failed:
Failed Test Cases:

real	0m18.937s
user	0m18.284s
sys	0m0.596s
```
(Time of implementation with `unordered_map` and `absl::Hash`)

I've also tested with `absl::flat_hash_map` but the results were slightly worse.

Reply:
Nice! I just discovered that the [coding guidelines](https://github.com/microsoft/onnxruntime/blob/1b07bbceaa234a37d0418d8e61f2d52dbd80af37/docs/Coding_Conventions_and_Standards.md#containers-to-use) state that we should not depend on `absl` directly, though :( I'm sorry for sending you around the block! @xadupre how would you proceed?

Reply:
The guidelines should be followed, so no absl. I don't think the hash we used is an issue (std::hash is used). I did not check the algorithm but that should have a uniform distribution. You used 2**60  in your experiments, so I assumed you have 2** 60 nodes. Then 2 ** 60 = 2 ** 32 * 2 ** 28 = 2 ** sizeof(uint32_t) * average(number of collisions). My guess is no hash can reduce the number of collisions unless it uses something bigger than uint32_t to store it, so uint64_t. Since this map is used to build the internal structure and disappears after that, I would try to a hash function like `std::hash<uint64_t>(static_cast<uint64_t>(id_tree) << 32 | static_cast<uint64_t>(id_node))`. Would that be ok for you?

Reply:
We have 3 million nodes spread through 5 thousand trees. I used 2^60 just to show that the proposed hash function would behave well even for large ID's that would cause it to overflow.

The solution you proposed works well for us. I proposed Elegant pairing as it's a more generic approach capable of handling different ID allocation strategies since we have few guarantees regarding it. My understanding is we can count only on having each pair of ID's (tree and node) unique within `int64_t` x `int64_t` range. In other words, since it would reduce any possible impacts, I thought it would help approving this PR 🙂 .

But I agree assuming 32 bits from hashing perspective is reasonable since it seems converters start allocation of ID's from 1 and few models would have more than 4 billion nodes. I will proceed with the required changes.

Reply:
FYI, the std hash function does not (necessarily) have a uniform distribution. It seems to be a no-op in our case: https://godbolt.org/z/4675qPeYd 

Either way, I was hoping there was a quick and simple way that Just Works. In its absence, I think either the current or elegant pairing is just fine :)

Reply:
It seems std::hash is optimized for unordered_map and std::hash<int>(i) returns i. We can even remove it in this case.

Reply:
Changed accordingly 🙂

Reply:
Unresolved for future readers